File: src/main/java/nz/co/searchwellington/filters/attributesetters/FeedAttributeSetter.java
package nz.co.searchwellington.filters.attributesetters;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.servlet.http.HttpServletRequest;

import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.repositories.HibernateResourceDAO;

import org.apache.log4j.Logger;

public class FeedAttributeSetter implements AttributeSetter {
	
	private static Logger log = Logger.getLogger(FeedAttributeSetter.class);
	
	public static final String FEED_ATTRIBUTE = "feedAttribute";

	private static Pattern feedPattern = Pattern.compile("^/feed/(.*?)(/(edit|save|rss|json))?$");

	private HibernateResourceDAO resourceDAO;
	
	public FeedAttributeSetter(HibernateResourceDAO resourceDAO) {
		this.resourceDAO = resourceDAO;
	}
	
	@Override
	public boolean setAttributes(HttpServletRequest request) {
		log.debug("Looking for feed path");
		Matcher contentMatcher = feedPattern.matcher(request.getPathInfo());
		if (contentMatcher.matches()) {
			final String match = contentMatcher.group(1);
			log.debug("'" + match + "' matches content");
			log.debug("Looking for feed '" + match + "'");
			Feed feed = (Feed) resourceDAO.loadFeedByUrlWords(match);
			if (feed != null) {
				log.info("Setting feed: " + feed.getName());
				request.setAttribute(FEED_ATTRIBUTE, feed);
				request.setAttribute("resource", feed);
				return true;
			}
		}
		return false;
	}
	
}

File: src/main/scala/nz/co/searchwellington/filters/attributesetters/FeedAttributeSetter.scala
package nz.co.searchwellington.filters.attributesetters

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.filters.RequestPath
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import java.util.regex.Pattern
import javax.servlet.http.HttpServletRequest
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

@Component
class FeedAttributeSetter @Autowired()(mongoRepository: MongoRepository) extends AttributeSetter with ReasonableWaits {

  private val feedPattern = Pattern.compile("^/feed/(.*?)(/(edit|save|rss|json|accept-all))?$")

  override def setAttributes(request: HttpServletRequest): Future[Map[String, Any]] = {
    val contentMatcher = feedPattern.matcher(RequestPath.getPathFrom(request))
    if (contentMatcher.matches) {
      val urlWords = contentMatcher.group(1)
      mongoRepository.getFeedByUrlwords(urlWords).map { maybeFeed =>
          maybeFeed.map { feed =>
            Map (
              FeedAttributeSetter.FEED_ATTRIBUTE -> feed,
              "resource" -> feed
            )
          }.getOrElse{
            Map.empty
          }
      }
    } else {
      Future.successful(Map.empty)
    }
  }
}

object FeedAttributeSetter {
  val FEED_ATTRIBUTE = "feedAttribute"
}
